### ISSUE TITLE
False positive: Operator "&=" not supported for types "Never" and "C"

### ISSUE THREAD

**Issue by dimaqq** [microsoft/pyright#2283]
```py
# integers work


def intfoo(bar: int):
    rv = None
    if bar % 2:
        if rv:
            rv &= bar
        else:
            rv = bar


# but custom classes don't seem to?

class C:
    def __and__(self, other: "C"):
        return C()

def customfoo(bar: C):
    rv = None
    if bar:
        if rv:
            rv &= bar
        else:
            rv = bar
```

Somehow the `int` example is OK by `pyright`.
But the `C` example is not 😕 

Output:
```command
> npx pyright test-pyright.py
Found 1 source file
/path/to/test-pyright.py
  /path/to/test-pyright.py:23:13 - error: Operator "&=" not supported for types "Never" and "C"
    Operator "&" not supported for types "Never" and "C" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 infos
```

Tested with `pyright 1.1.166`

**Comment by erictraut** [comment]
You are setting `rv` to the value of `None` and then using it in the conditional statement `if rv:`. This is performing type narrowing on `rv`, and it narrows it to type `Never` because `if None` is always false. In other words, this code path will never be taken.

**Comment by erictraut** [comment]
The reason the `int` case works is because the `int` class implements the `__rand__` operator. If you implement this as well in class `C`, the error will go away.

**Comment by dimaqq** [comment]
Interesting!

I guess `bar.__rand__(<Never>)` should also trigger a pyright error, right?

**Comment by dimaqq** [comment]
Maybe slightly more to the point: "code unreachable" would be a better error in this case 🤔

**Comment by erictraut** [comment]
`bar.__rand__(<Never>)` will not trigger an error if your `__rand__` function accepts an `Any` parameter because all types (even `Never`) are considered assignable to `Any`.

**Comment by dimaqq** [comment]
re: ⬆️
The error is not triggered in this case, without `Any`:

```py
class C:
    def __and__(self, other: "C"):
        return C()
    def __rand__(self, other: "C"):
        return C()

def customfoo(bar: C):
    rv = None
    if bar:
        if rv:
            rv &= bar
        else:
            rv = bar
```
